src/Voron/Data/Tables/TableValueReader.cs

```diff
@@ -54,65 +116,87 @@ public long ReadLong(int index)
+            return Encoding.UTF8.GetString(read, size);
+        }
+
         public string ReadStringWithPrefix(int index, int bytesToSkip)
         {
             byte* read = Read(index, out var size);
             return Encoding.UTF8.GetString(read + bytesToSkip, size - bytesToSkip);
         }
 
+        public string ReadStringWithPrefix<TConstant>(int index, int bytesToSkip) where TConstant : struct, INumericConstant
+        {
+            byte* read = Read<TConstant>(index, out var size);
+            return Encoding.UTF8.GetString(read + bytesToSkip, size - bytesToSkip);
+        }
+
         // PERF: This table allow us to access the <mask, shift, indexAdjustment> tuple directly from a location in memory.
         // The reason why we use an extra integer is to avoid having to multiply by 3, by multiplying by 4 we can
         // multiply by applying shifts (the JIT will take care of that).
         private static ReadOnlySpan<int> ElementsTable => new int[] 
-            {
-                0, 0, 0, 0,                             // Index 0 (not used)
-                0x000000FF, 24, 3, 0,                   // Index 1 when _elementSize == 1
-                0x0000FFFF, 16, 2, 0,                   // Index 2 when _elementSize == 2
-                0, 0, 0, 0,                             // Index 3 (not used)
-                unchecked((int)0xFFFFFFFF), 0, 0, 0     // Index 4 when _elementSize == 4
-            };
-        
-        public byte* Read(int index, out int size)
+        {
+            0, 0, 0, 0,                             // Index 0 (not used)
+            0x000000FF, 24, 3, 0,                   // Index 1 when _elementSize == 1
+            0x0000FFFF, 16, 2, 0,                   // Index 2 when _elementSize == 2
+            0, 0, 0, 0,                             // Index 3 (not used)
+            unchecked((int)0xFFFFFFFF), 0, 0, 0     // Index 4 when _elementSize == 4
+        };
+
+        internal byte* Read<TConstant>(int index, out int size) where TConstant : struct, INumericConstant
         {
             PortableExceptions.ThrowIfOnDebug<ArgumentOutOfRangeException>(index < 0 || index >= Count);
 
             // Calculate the starting pointer to the data, adjusted by the offset.
             var dataPtr = (Pointer + _offset);
 
-            // Retrieve mask, shift, and index adjustment values from the table.
-            int elementsLoc = 4 * _elementSize;
-            uint mask = (uint)ElementsTable[elementsLoc];
-            int shift = ElementsTable[elementsLoc + 1];
-            int indexAdjustment = ElementsTable[elementsLoc + 2];
+            // Retrieve mask, shift, and index adjustment values from the table and bake all of them 
+            // on into a constant. 
+            uint mask = (uint)ElementsTable[4 * default(TConstant).N];
+            int shift = ElementsTable[4 * default(TConstant).N + 1];
+            int indexAdjustment = ElementsTable[4 * default(TConstant).N + 2];
```

Comment:
Does it mean that the primary motivation for all / most of the work made in this commit is to have _constant_ values here?

Reply:
Yes. The idea is that now all those becomes constants and those constants get pushed downstream and allow much better and compact code.